driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java

```diff
@@ -154,12 +159,21 @@ public void abortTransaction() {
                         readConcern, this);
             }
         } catch (Exception e) {
-            // ignore errors
+            if (e instanceof MongoException) {
```

Comment:
I suspect that once you deal with unpinning of mongos in CommitOperation and AbortOperation, you'll be able to remove all the unpinning logic in this class.